cin_validator/cin_validator_class.py

```diff
@@ -0,0 +1,109 @@
+                issue_dfs_per_rule = pd.Series(
+                    [
+                        ctx.type_zero_issues,
+                        ctx.type_one_issues,
+                        ctx.type_two_issues,
+                        ctx.type_three_issues,
+                    ]
+                )
+                # error_df_lengths is a list of lengths of all elements in issue_dfs_per_rule respectively.
+                error_df_lengths = pd.Series([len(x) for x in issue_dfs_per_rule])
+                if error_df_lengths.max() == 0:
+                    # if the rule didn't push to any of the issue accumulators, then it didn't find any issues in the file.
+                    self.rules_passed.append(rule.code)
+                else:
+                    # get the rule type based on which attribute had elements pushed to it (i.e non-zero length)
+                    # its corresponding error_df can be found by issue_dfs_per_rule[ind]
+                    ind = error_df_lengths.idxmax()
+
+                    issue_dict = {
+                        "code": rule.code,
+                        "number": error_df_lengths[ind],
+                        "type": ind,
+                    }
+                    issue_dict_df = pd.DataFrame([issue_dict])
+                    self.issue_instances = pd.concat(
+                        [self.issue_instances, issue_dict_df], ignore_index=True
+                    )
+
+                    # add a the rule's code to it's error_df
+                    issue_dfs_per_rule[ind]["rule_code"] = rule.code
+
+                    # temporary: add rule type to track if all types are in df.
+                    issue_dfs_per_rule[ind]["rule_type"] = ind
+
+                    # combine this rule's error_df with the cummulative error_df
+                    self.all_rules_issue_locs = pd.concat(
+                        [self.all_rules_issue_locs, issue_dfs_per_rule[ind]],
+                        ignore_index=True,
+                    )
+
+            except Exception as e:
+                print(f"Error with rule {rule.code}: {type(e).__name__}, {e}")
+
+    def create_json_report(self):
+        self.json_issue_report = self.all_rules_issue_locs.to_dict(orient="records")
+
+    def select_by_id(self):
+        if self.issue_id is not None:
+            print("reached hereeeeeeeeeeeeeeeee")
+            self.issue_id = tuple(self.issue_id.split(", "))
+            self.all_rules_issue_locs = self.all_rules_issue_locs[
+                self.all_rules_issue_locs["ERROR_ID"] == self.issue_id
+            ]
+        else:
+            pass
+
+    def create_rules_df(self):
+        # rules_broken = self.all_rules_issue_locs
+        rules_broken = (
+            self.all_rules_issue_locs["rule_code"].astype("str").unique().tolist()
+        )
+        print(rules_broken)
+        rule_messages = []
+        for rule in registry:
+            if str(rule.code) in rules_broken:
```

Comment:
Hi Will, this works does work.

On second thought however, let us not loop through registry twice (we are doing that in the create_error_report_df method).
How about merging this feature with the create_error_report_df function such that when we identify the rules that fail, we immediately add the rule code-description pair to a  dictionary as you did with issue_dict?

What do you think? @willLPD2I

Reply:
@tab1tha Sounds way better! Cool!